resources/net.go

```diff
@@ -0,0 +1,849 @@
+					return
+				}
+				select {
+				case nlChan <- &nlChanStruct{
+					err: errwrap.Wrapf(err, "error selecting on fd"),
+				}:
+				case <-closeChan:
+				}
+				return
+			}
+			// receive the message from the netlink socket into b
+			b := make([]byte, os.Getpagesize())
+			n, _, err := unix.Recvfrom(nlSock, b, unix.MSG_DONTWAIT) // non-blocking receive
+			if err != nil {
+				// if nlSock is closed
+				if err == unix.EBADF { // bad file descriptor
+					return
+				}
+				select {
+				case nlChan <- &nlChanStruct{
+					err: errwrap.Wrapf(err, "error receiving messages"),
+				}:
+				case <-closeChan:
+				}
+				return
+			}
+			// if we didn't get enough bytes for a header, something went wrong
+			if n < unix.NLMSG_HDRLEN {
+				select {
+				case nlChan <- &nlChanStruct{
+					err: fmt.Errorf("received short header"),
+				}:
+				case <-closeChan:
+				}
+				return
+			}
+			b = b[:n] // truncate b to message length
+			// use syscall to parse, as func does not exist in x/sys/unix
+			msgs, err := syscall.ParseNetlinkMessage(b)
+			if err != nil {
+				select {
+				case nlChan <- &nlChanStruct{
+					err: errwrap.Wrapf(err, "error parsing netlink message"),
+				}:
+				case <-closeChan:
+					return
+				}
+			}
+			select {
+			case nlChan <- &nlChanStruct{
+				msg: msgs,
+			}:
+			case <-closeChan:
+				return
+			}
+		}
+	}()
+
+	// notify engine that we're running
+	if err := obj.Running(); err != nil {
+		return err // bubble up a NACK...
+	}
+
+	var exit *error
+	var send bool
+	var done bool
+
+	for {
+		select {
+		case s, ok := <-nlChan:
+			if !ok {
+				if done {
+					return nil
+				}
+				done = true
+				continue
+			}
+			if err := s.err; err != nil {
+				return errwrap.Wrapf(s.err, "unknown netlink error")
+			}
+			if obj.debug {
+				log.Printf("%s: Event: %+v", obj, s.msg)
+			}
+
+			send = true
+			obj.StateOK(false)
+
+		case event, ok := <-recWatcher.Events():
+			if !ok {
+				if done {
+					return nil
+				}
+				done = true
+				continue
+			}
+			if err := event.Error; err != nil {
+				return errwrap.Wrapf(err, "unknown recwatcher error")
+			}
+			if obj.debug {
+				log.Printf("%s: Event(%s): %v", obj, event.Body.Name, event.Body.Op)
+			}
+
+			send = true
+			obj.StateOK(false) // dirty
+
+		case event := <-obj.Events():
+			if exit, send = obj.ReadEvent(event); exit != nil {
+				return *exit // exit
+			}
+		}
+
+		// do all our event sending all together to avoid duplicate msgs
+		if send {
+			send = false
+			obj.Event()
+		}
+	}
+}
+
+// ifaceCheckApply checks the state of the network device and brings it up or
+// down as necessary.
+func (obj *NetRes) ifaceCheckApply(apply bool) (bool, error) {
+	// check the interface state
+	state, err := obj.iface.state()
+	if err != nil {
+		return false, errwrap.Wrapf(err, "error checking %s state", obj.GetName())
+	}
+	// if up is unspecified
+	if obj.State == "" {
+		obj.State = state // send the state before returning
```

Comment:
This is incorrect. Do not send state. If state was unspecified it means "we're not touching it", but now we've said "make sure it's up". Easy one line fix though :)

Logic below will change though.